### PR TITLE
Fix a memory leak in AlignedVector.

### LINKED ISSUE #12998: Memory leak `Table`/`AlignedVector`
as pointed out in #12993

### PR DESCRIPTION
**Pull request by bangerth**
Fixes #12998. As pointed out in #12993.

/rebuild

### CODE REVIEW

File: include/deal.II/base/aligned_vector.h [hunk 1]
```diff
@@ -1199,11 +1199,22 @@ template <class T>
 inline AlignedVector<T> &
 AlignedVector<T>::operator=(const AlignedVector<T> &vec)
 {
+  const size_type new_size = vec.used_elements_end - vec.elements.get();
+
+  // First throw away everything and re-allocate memory but leave that
+  // memory uninitialized for now:
   resize(0);
-  resize_fast(vec.used_elements_end - vec.elements.get());
+  reserve(new_size);
```
Comment (bangerth) [inline include/deal.II/base/aligned_vector.h:1207]: The bug is here: When we call `resize_fast()`, we allocate memory and default construct objects in that memory. But in the line below (`AlignedVectorCopyConstruct`) we copy-construct *other* objects in that same memory location, forgetting to call the destructor on what we had done before. 

The solution is to not initialize the memory we allocate -- that's what `reserve()` does in contrast to `resize*()`.

Comment (bangerth) [reply include/deal.II/base/aligned_vector.h:1207]: For reference, on second look, I see that in essence the same code I just wrote here is also what's in the copy constructor just a few lines up.

**Comment by bangerth** [comment]
@elauksap FYI.

**Review by masterleinad — APPROVED** [review]

**Comment by elauksap** [comment]
> @elauksap FYI.

That's brilliant, thanks a lot!

**Review by kronbichler — APPROVED** [review]

**Comment by peterrum** [comment]
@elauksap Could you try out this patch without #12995.

**Comment by elauksap** [comment]
@peterrum just tried and it works! You can safely revert [my patch](530823a868072f92fc440541e388b53c90da3528), if needed.

**Review by peterrum — APPROVED** [review]

**Comment by kronbichler** [comment]
I would vote for keeping the patch, it is a nice use of compile-time knowledge and hence more direct in my opinion.

**Pull request merged**